### ISSUE TITLE
Question about docker-compose (or dependency dockers)

### ISSUE THREAD

**Issue by Noodle05** [KostyaSha/yet-another-docker-plugin#118]
Our product will depend on bunch of services like zookeeper for example. So in order to run integration test, it will require some other docker containers.

The plan is: use something like docker-compose to define group of containers and create separate network for each test (docker-compose will create network with project name as prefix). And start jenkins slave docker attach to this network to run test.

So, my question is: how can I launch a docker slave with specific network attached?

Thanks

**Comment by KostyaSha** [comment]
If there is no such option in container create options, then could be possible to add.
Docker compose - i bit lost what docker added in last versions, i have swarm support in docker-java (and here) on high priority. if there is something that may work via API, then compose is possible.

**Comment by Noodle05** [comment]
Docker compose is not swarm. Docker compose is better to handle/manage group of micro services to provide service and better to run multiple service instances on same docker cloud.

For example: I can create a single docker compose yml file, and start multiple instance by providing different project name. Internally, docker-compose will create network, volume, and containers with project name as prefix. So I can have two instance of docker-compose service running at same time. 

Of cause, port mapping need to be taken care, but since this is integration test, I don't really need to export any port.

I'm thinking: in a test, use shell script to call docker-compose to start required contains, get network name this docker-compose instance created, and create slave docker with "--net <docker-compose network name", run test in slave, and then tear down by remove slave docker as well as docker-compose down -v to clean up. So each test will have their own environment.

With current implementation, is it possible?

BTW, thank you for rapid response. Very appreciate.

**Comment by KostyaSha** [comment]
Dublicates https://github.com/KostyaSha/yet-another-docker-plugin/issues/45 ?

**Comment by Noodle05** [comment]
Probably duplicated, but is there a way to create slave docker with "--net" option?

**Comment by KostyaSha** [comment]
> Probably duplicated, but is there a way to create slave docker with "--net" option?

https://github.com/KostyaSha/yet-another-docker-plugin/blob/master/yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/commons/DockerCreateContainer.java#L522 ?

**Comment by KostyaSha** [comment]
https://github.com/KostyaSha/yet-another-docker-plugin/blob/master/yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/commons/DockerCreateContainer/help-networkMode.html or it something else?

Please point on https://docs.docker.com/engine/reference/api/docker_remote_api/ option

**Comment by Noodle05** [comment]
That's only the network mode, not specific network?

**Comment by Noodle05** [comment]
Didn't found full syntax on create container, but find API to attach network:

POST /networks/(id)/connect now allows you to set the static IPv4 and/or IPv6 address for the container.

**Comment by Noodle05** [comment]
Here it is:

https://docs.docker.com/engine/reference/api/docker_remote_api_v1.23/#/create-a-container

It contains network section, which specify which network to attach, alias name and so on.

**Comment by KostyaSha** [comment]
How it correlates with `"NetworkMode": "bridge",` that already exist ?

**Comment by Noodle05** [comment]
After I create a container using docker command and then docker inspect, the "NetworkMode" shows as the network it attached to.

**Comment by Noodle05** [comment]
NetworkMode - Sets the networking mode for the container. Supported standard values are: bridge, host, none, and container:<name|id>. Any other value is taken as a custom network’s name to which this container should connect to.

So, "NetworkMode" is custom network's name.

Now the question is: how to specify it when create slave docker (instead of define in template?)

**Comment by KostyaSha** [comment]
it must be defined in template, template is configuration. First of all docker-java need API, then i use this API in this plugin. Will try add when will have time.

**Comment by Noodle05** [comment]
In that case, can network mode use variable like ${project_name}-net in template?

**Comment by KostyaSha** [comment]
Standard cloud api runs slaves for labels, jenkins doesn't provide job context.
Why not simply run docker-compose from job as build step?

**Comment by Noodle05** [comment]
Do you think this will work?

1. Create an yad cloud.
2. Create multiple template, each with an unique network mode (customized network), and accept only one job.
3. Within slave docker, use shell script run docker-compose to start dependent docker containers with customized network (use label as network name maybe?).
4. Continue regular build, test in slave docker.

If this work, we need 
1. create docker template for each docker network/slave
2. Install docker client and docker-compose in slave docker image.

You see simply run docker-compose is not scalable and flexible, so it's not simply.

Unless we can run jenkins's slave.jar in an existing docker container (that created by docker-compose) without use docker plugin? In that case, we can still use jenkin's function (like issue build command, test command, get code coverage report and so on)

Or my understanding of slave is wrong?

**Comment by KostyaSha** [comment]
> Do you think this will work?
> 
> Create an yad cloud.
> Create multiple template, each with an unique network mode (customized network), and accept only one job.
> Within slave docker, use shell script run docker-compose to start dependent docker containers with customized network (use label as network name maybe?).
> Continue regular build, test in slave docker.
> If this work, we need
> 
> create docker template for each docker network/slave
> Install docker client and docker-compose in slave docker image.

yes. This is top usage case, plugin starts slave in container with any tools and then it do everything needed for testing, in the end kills.


> You see simply run docker-compose is not scalable and flexible, so it's not simply.

yes


> Unless we can run jenkins's slave.jar in an existing docker container (that created by docker-compose) without use docker plugin? In that case, we can still use jenkin's function (like issue build command, test command, get code coverage report and so on)

Of course you can run slave.jar in any way.
